Error: Cannot find module '/package.json'

**Platform:**
Linux pop-os 5.15.11-76051511-generic #202112220937~1640185481~21.04~b3a2c21-Ubuntu SMP Mon Jan 3 16:5 x86_64 x86_64 x86_64 GNU/Linux

**Nuclear version:**
nuclear v0.6.17 
installed from  nuclear-v0.6.17.deb

**Description of the issue:**

nuclear opens but is unable to play anything.


nuclear
      main › (node:8306) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `nuclear --trace-deprecation ...` to show where the warning was created)
      main › (node:8306) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'write' of null
    at e.exports.send (/opt/nuclear/resources/app.asar/dist/main.js:2200:13798)
    at /opt/nuclear/resources/app.asar/dist/main.js:2200:2154
    at new Promise (<anonymous>)
    at e.exports.request (/opt/nuclear/resources/app.asar/dist/main.js:2200:2107)
    at e.exports.clearActivity (/opt/nuclear/resources/app.asar/dist/main.js:2200:6785)
    at fr.clear (/opt/nuclear/resources/app.asar/dist/main.js:2751:510734)
    at ca.onClearTrackList (/opt/nuclear/resources/app.asar/dist/main.js:2751:531161)
    at IpcMainImpl.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:536446)
    at IpcMainImpl.emit (events.js:315:20)
    at Object.<anonymous> (electron/js2c/browser_init.js:161:9715)
    at Object.emit (events.js:315:20)
      main › (node:8306) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
      main › (node:8306) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
-----

    main › (node:8306) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `nuclear --trace-deprecation ...` to show where the warning was created)
      main › (node:8306) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'write' of null
    at e.exports.send (/opt/nuclear/resources/app.asar/dist/main.js:2200:13798)
    at /opt/nuclear/resources/app.asar/dist/main.js:2200:2154
    at new Promise (<anonymous>)
    at e.exports.request (/opt/nuclear/resources/app.asar/dist/main.js:2200:2107)
    at e.exports.clearActivity (/opt/nuclear/resources/app.asar/dist/main.js:2200:6785)
    at fr.clear (/opt/nuclear/resources/app.asar/dist/main.js:2751:510734)
    at ca.onClearTrackList (/opt/nuclear/resources/app.asar/dist/main.js:2751:531161)
    at IpcMainImpl.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:536446)
    at IpcMainImpl.emit (events.js:315:20)
    at Object.<anonymous> (electron/js2c/browser_init.js:161:9715)
    at Object.emit (events.js:315:20)
      main › (node:8306) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
      main › (node:8306) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.


nuclear --trace-deprecation 
      main › (node:9092) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `nuclear --trace-deprecation ...` to show where the warning was created)
  renderer › Error: Cannot find module '/package.json'
    at n (file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:162754)
    at M (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:83672)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:84166)
    at file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80126
    at Array.map (<anonymous>)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80119)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Object.t.liveStreamSearch (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:168100)
    at async file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:215725
    at async Promise.all (index 5)
  renderer › Error: Cannot find module '/package.json'
    at n (file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:162754)
    at M (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:83672)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:84166)
    at file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80126
    at Array.map (<anonymous>)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80119)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Object.t.liveStreamSearch (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:168100)
    at async file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:215725
    at async Promise.all (index 5)
      main › (node:9092) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'write' of null
    at e.exports.send (/opt/nuclear/resources/app.asar/dist/main.js:2200:13798)
    at /opt/nuclear/resources/app.asar/dist/main.js:2200:2154
    at new Promise (<anonymous>)
    at e.exports.request (/opt/nuclear/resources/app.asar/dist/main.js:2200:2107)
    at e.exports.clearActivity (/opt/nuclear/resources/app.asar/dist/main.js:2200:6785)
    at fr.clear (/opt/nuclear/resources/app.asar/dist/main.js:2751:510734)
    at ca.onClearTrackList (/opt/nuclear/resources/app.asar/dist/main.js:2751:531161)
    at IpcMainImpl.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:536446)
    at IpcMainImpl.emit (events.js:315:20)
    at Object.<anonymous> (electron/js2c/browser_init.js:161:9715)
    at Object.emit (events.js:315:20)
      main › (node:9092) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
      main › (node:9092) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
  renderer › Error while searching  for The O'Jays Back Stabbers on Youtube
  renderer › Error: Cannot find module '/package.json'
    at n (file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:162754)
    at M (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:83672)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:84166)
    at file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80126
    at Array.map (<anonymous>)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80119)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async N (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:166608)
    at async c.search (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:147218)
    at async file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:207770
    at async file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:207486
      main › (node:9092) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'write' of null
    at e.exports.send (/opt/nuclear/resources/app.asar/dist/main.js:2200:13798)
    at /opt/nuclear/resources/app.asar/dist/main.js:2200:2154
    at new Promise (<anonymous>)
    at e.exports.request (/opt/nuclear/resources/app.asar/dist/main.js:2200:2107)
    at e.exports.clearActivity (/opt/nuclear/resources/app.asar/dist/main.js:2200:6785)
    at fr.clear (/opt/nuclear/resources/app.asar/dist/main.js:2751:510734)
    at App.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:537783)
    at App.emit (events.js:327:22)
    at Lt.close (/opt/nuclear/resources/app.asar/dist/main.js:2751:486639)
    at ha.onClose (/opt/nuclear/resources/app.asar/dist/main.js:2751:533655)
    at IpcMainImpl.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:536446)
    at IpcMainImpl.emit (events.js:315:20)
    at Object.<anonymous> (electron/js2c/browser_init.js:161:9715)
    at Object.emit (events.js:315:20)
      main › (node:9092) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)

-----

     main › (node:9385) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `nuclear --trace-deprecation ...` to show where the warning was created)
  renderer › Error: Cannot find module '/package.json'
    at n (file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:162754)
    at M (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:83672)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:84166)
    at file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80126
    at Array.map (<anonymous>)
    at e.exports (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:80119)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Object.t.liveStreamSearch (file:///opt/nuclear/resources/app.asar/dist/1.renderer.js:2685:168100)
    at async file:///opt/nuclear/resources/app.asar/dist/renderer.js:1:215725
    at async Promise.all (index 5)
      main › error trying to set discord activity
      main › (node:9385) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'write' of null
    at e.exports.send (/opt/nuclear/resources/app.asar/dist/main.js:2200:13798)
    at /opt/nuclear/resources/app.asar/dist/main.js:2200:2154
    at new Promise (<anonymous>)
    at e.exports.request (/opt/nuclear/resources/app.asar/dist/main.js:2200:2107)
    at e.exports.clearActivity (/opt/nuclear/resources/app.asar/dist/main.js:2200:6785)
    at fr.clear (/opt/nuclear/resources/app.asar/dist/main.js:2751:510734)
    at App.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:537783)
    at App.emit (events.js:327:22)
    at Lt.close (/opt/nuclear/resources/app.asar/dist/main.js:2751:486639)
    at ha.onClose (/opt/nuclear/resources/app.asar/dist/main.js:2751:533655)
    at IpcMainImpl.<anonymous> (/opt/nuclear/resources/app.asar/dist/main.js:2751:536446)
    at IpcMainImpl.emit (events.js:315:20)
    at Object.<anonymous> (electron/js2c/browser_init.js:161:9715)
    at Object.emit (events.js:315:20)
      main › (node:9385) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
      main › (node:9385) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
   ipc api › error in event start-download => Object has been destroyed
   ipc api › TypeError: Object has been destroyed
    at Lt.send (/opt/nuclear/resources/app.asar/dist/main.js:2751:485660)
    at Qt.onStartDownload (/opt/nuclear/resources/app.asar/dist/main.js:2751:489414)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

--- a/packages/main/__mocks__/@nuclear/core.ts
+++ b/packages/main/__mocks__/@nuclear/core.ts
@@ -1,0 +1,70 @@
+export default jest.fn();
+export const settingsConfig = jest.requireActual('@nuclear/core/src/settings').settingsConfig;
+
+export enum IpcEvents {
+    STARTED = 'started',
+  
+    PLAY = 'play',
+    PAUSE = 'pause',
+    STOP = 'stop',
+    PLAYPAUSE = 'playpause',
+    NEXT = 'next',
+    PREVIOUS = 'previous',
+    VOLUME = 'volume',
+    MUTE = 'mute',
+    SEEK = 'seek',
+    PLAYING_STATUS = 'playing-status',
+    SONG_CHANGE = 'song-change',
+  
+    LOCALFOLDER_REMOVE = 'remove-localfolder',
+    LOCALFOLDERS_REFRESH = 'refresh-localfolders',
+    LOCALFOLDERS_GET = 'get-localfolders',
+    LOCALFOLDERS_SET = 'set-localfolders',
+    LOCAL_METAS = 'get-metas',
+  
+    TRACK_ADD = 'add-track',
+    TRACK_REMOVE = 'remove-track',
+    TRACK_SELECT = 'select-track',
+    QUEUE_CLEAR = 'clear-queue',
+    QUEUE = 'queue',
+    QUEUE_ADD = 'queue-add',
+    QUEUE_DROP = 'queue_drop',
+  
+    SETTINGS = 'settings',
+    SHUFFLE = 'shuffle',
+    LOOP = 'loop',
+  
+    CONNECTIVITY = 'connectivity',
+  
+    DOWNLOAD_START = 'start-download',
+    DOWNLOAD_PAUSE = 'pause-download',
+    DOWNLOAD_STARTED = 'download-started',
+    DOWNLOAD_PROGRESS = 'download-progress',
+    DOWNLOAD_FINISHED = 'download-finished',
+    DOWNLOAD_ERROR = 'download-error',
+    DOWNLOAD_REMOVED = 'download-removed',
+  
+    WINDOW_MINIMIZE = 'minimize',
+    WINDOW_MAXIMIZE = 'maximize',
+    WINDOW_CLOSE = 'close',
+    WINDOW_OPEN_DEVTOOLS = 'open-devtools',
+  
+    API_RESTART = 'restart-api',
+    API_STOP = 'stop-api',
+  
+    PLAYLIST_CREATE = 'create-playlist',
+    PLAYLIST_REFRESH = 'refresh-playlist',
+    PLAYLIST_ACTIVATE = 'activate-playlist',
+    PLAYLIST_ADD_QUEUE = 'add-queue-playlist',
+  
+    EQUALIZER_UPDATE = 'update-equalizer',
+    EQUALIZER_SET = 'set-equalizer',
+  
+    LOCAL_FILES = 'local-files',
+    LOCAL_FILES_PROGRESS = 'local-files-progress',
+    LOCAL_FILES_ERROR = 'local-files-error',
+  
+    PLAY_STARTUP_TRACK = 'play-startup-track',
+  
+    ELECTRON_TIMBER_ERROR_EVENT = '__ELECTRON_TIMBER_ERROR__'
+  }

--- a/packages/main/__mocks__/electron-devtools-installer.ts
+++ b/packages/main/__mocks__/electron-devtools-installer.ts
@@ -1,0 +1,1 @@
+export default jest.fn();

--- a/packages/main/__mocks__/electron-store.ts
+++ b/packages/main/__mocks__/electron-store.ts
@@ -1,0 +1,6 @@
+export function get() {
+  return '';
+}
+export function set() {}
+
+export default class ElectronStore {}

--- a/packages/main/__mocks__/electron-timber.ts
+++ b/packages/main/__mocks__/electron-timber.ts
@@ -1,0 +1,3 @@
+export default {
+  hookConsole: jest.fn()
+};

--- a/packages/main/__mocks__/electron.ts
+++ b/packages/main/__mocks__/electron.ts
@@ -1,0 +1,3 @@
+export const app = {
+  getPath: () => ''
+};

--- a/packages/main/jest.config.ts
+++ b/packages/main/jest.config.ts
@@ -1,0 +1,5 @@
+export default async () => ({
+  transform: {
+    '^.+\\.(ts|tsx|js|jsx)$': 'ts-jest'
+  }
+});

--- a/packages/main/package.json
+++ b/packages/main/package.json
@@ -6,7 +6,8 @@
     "start": "shx rm -rf build && webpack --env.NODE_ENV=development && electron build/main.js --inspect=9001",
     "build": "cross-env NODE_OPTIONS=--max-old-space-size=8192 webpack --colors --env.NODE_ENV=production",
     "lint": "eslint src/**/**.ts",
-    "lint:fix": "eslint src/**/**.ts --fix"
+    "lint:fix": "eslint src/**/**.ts --fix",
+    "test": "jest"
   },
   "repository": {
     "type": "git",
@@ -65,6 +66,7 @@
     "@types/electron-devtools-installer": "^2.2.0",
     "@types/express": "^4.17.2",
     "@types/glob": "^7.1.1",
+    "@types/jest": "^27.4.0",
     "@types/lodash": "^4.14.158",
     "@types/node": "^12.12.17",
     "@types/node-fetch": "^2.5.4",
@@ -76,9 +78,11 @@
     "cross-env": "^7.0.3",
     "electron-builder": "^21.2.0",
     "eslint": "^7.18.0",
+    "jest": "^27.4.7",
     "node-gyp": "^7.0.0",
     "node-loader": "^0.6.0",
     "shx": "^0.3.3",
+    "ts-jest": "^27.1.3",
     "ts-loader": "^8.0.3",
     "ts-node": "^8.10.2",
     "type-fest": "^0.8.1",

--- a/packages/main/src/controllers/player.test.ts
+++ b/packages/main/src/controllers/player.test.ts
@@ -1,0 +1,42 @@
+import 'reflect-metadata';
+import IpcPlayer from './player';
+
+import Discord from '../services/discord';
+import Config from '../services/config';
+
+jest.mock('../services/discord', () => {
+  return function () {
+    return {
+      clear: jest.fn().mockRejectedValue(new Error('discord failed to clear'))
+    };
+  };
+});
+
+jest.mock('../services/logger');
+
+describe('IPC Player controller', () => {
+  beforeAll(() => {
+    process.env.ACOUSTIC_ID_KEY = 'acoustic_id_key';
+    process.env.DISCORD_CLIENT_ID = 'discord_client_id';
+  });
+  it('catches the error if Discord service throws it on queue clear', async () => {
+    const systemApi = jest.fn() as any;
+    const logger = {
+      log: jest.fn(),
+      error: jest.fn()
+    } as any;
+
+    const ipc = new IpcPlayer(
+      new Discord(new Config(logger), null),
+      null,
+      systemApi,
+      null,
+      logger
+    );
+    await ipc.onClearTrackList();
+
+    return expect(logger.error).toHaveBeenCalledWith(
+      'Main process failed to react to IPC clear queue event'
+    );
+  });
+});

--- a/packages/main/src/controllers/player.ts
+++ b/packages/main/src/controllers/player.ts
@@ -8,6 +8,7 @@ import Window from '../services/window';
 import { ipcEvent, ipcController } from '../utils/decorators';
 import Discord from '../services/discord';
 import TrayMenu from '../services/trayMenu';
+import Logger, { $mainLogger } from '../services/logger';
 
 @ipcController()
 class IpcPlayer {
@@ -15,7 +16,8 @@ class IpcPlayer {
     @inject(Discord) private discord: Discord,
     @inject(TrayMenu) private trayMenu: TrayMenu,
     @inject(SystemApi) private systemApi: NuclearApi,
-    @inject(Window) private window: Window
+    @inject(Window) private window: Window,
+    @inject($mainLogger) private logger: Logger
   ) {}
 
   @ipcEvent(IpcEvents.PLAY)
@@ -58,10 +60,14 @@ class IpcPlayer {
   }
 
   @ipcEvent(IpcEvents.QUEUE_CLEAR)
-  onClearTrackList() {
-    this.systemApi.clearTrackList && this.systemApi.clearTrackList();
-    this.discord.clear();
-    this.trayMenu.update({isPlaying: false, track: null});
+  async onClearTrackList() {
+    try {
+      this.systemApi.clearTrackList && this.systemApi.clearTrackList();
+      await this.discord.clear();
+      this.trayMenu.update({isPlaying: false, track: null});
+    } catch (e) {
+      this.logger.error('Main process failed to react to IPC clear queue event');
+    }
   }
 
   @ipcEvent(IpcEvents.SONG_CHANGE)

--- a/packages/main/src/main.ts
+++ b/packages/main/src/main.ts
@@ -96,7 +96,7 @@ app.on('window-all-closed', async () => {
       await httpApi.close();
     }
 
-    discord.clear();
+    await discord.clear();
     await localDb.cleanUnusedThumbnail();
   } catch (err) {
     logger.error('Something failed during app close');

--- a/packages/main/src/services/discord/index.ts
+++ b/packages/main/src/services/discord/index.ts
@@ -90,11 +90,11 @@ class Discord {
     }
   }
 
-  clear() {
+  async clear() {
     delete this.activity;
     if (this.isReady) {
       this.logger.log('clear discord activity');
-      this.rpc.clearActivity();
+      await this.rpc.clearActivity();
     }
   }
 }

